### PR TITLE
satysfi: init at 2018-03-07

### PR DESCRIPTION
**Pull request by mt-caret**
###### Motivation for this change

A typesetting system.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by mt-caret** [comment]
ping @vbgl

**Comment by vbgl** [comment]
@GrahamcOfBorg build ocaml-ng.ocamlPackages_4_06.satysfi

**Comment by vbgl** [comment]
Since this is not an OCaml library, it does not belong to `pkgs/top-level/ocaml-packages.nix`, `ocamlPackages`,  nor `pkgs/development/ocaml-modules`.

**Comment by mt-caret** [comment]
Fixed.

**Comment by mt-caret** [comment]
@GrahamcOfBorg build satysfi

File: pkgs/tools/typesetting/satysfi/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{ stdenv, fetchzip, fetchFromGitHub, ocaml, ocamlbuild, findlib, menhir,
+  ppx_deriving, uutf, result, core_kernel, bitv, batteries, yojson, camlimages }:
```
Comment (vbgl) [inline pkgs/tools/typesetting/satysfi/default.nix]: Can you take `ocamlPackages` as one argument all libraries separately?

File: pkgs/tools/typesetting/satysfi/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      cp -r ${lm}/* lib-satysfi/dist/fonts/
+      cp -r ${lm-math}/otf/latinmodern-math.otf lib-satysfi/dist/fonts/
+      make install PREFIX=$out LIBDIR=$out/share/satysfi
+    '';
+
+    meta = with stdenv.lib; {
+      homepage = https://github.com/gfngfn/SATySFi;
+      description = "A statically-typed, functional typesetting system";
+      license = licenses.lgpl3;
+      maintainers = [ maintainers.mt-caret ];
```
Comment (vbgl) [inline pkgs/tools/typesetting/satysfi/default.nix]: Please add a `platforms` attribute.

File: pkgs/tools/typesetting/satysfi/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    src = fetchFromGitHub {
+      owner = "gfngfn";
+      repo = "SATySFi";
+      rev = "a050ec0906d083682c630b0dea68887415b5f53d";
+      sha256 = "12bhl7s2kc02amr8rm71pihj203f2j15y5j0kz3swgsw0gqh81gv";
+      fetchSubmodules = true;
+    };
+    buildInputs = [ findlib menhir ppx_deriving uutf result core_kernel bitv
+      batteries yojson camlimages ];
+    nativeBuildInputs = [ ocaml ocamlbuild ];
```
Comment (vbgl) [inline pkgs/tools/typesetting/satysfi/default.nix]: Why this split between `buildInputs` and `nativeBuildInputs`?

File: pkgs/tools/typesetting/satysfi/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+let
+  lm = fetchzip {
+    url = "http://www.gust.org.pl/projects/e-foundry/latin-modern/download/lm2.004otf.zip";
+    sha256 = "1mc88fbhfd2wki2vr700pgv96smya6d1z783xs3mfy138yb6ga2p";
+    stripRoot = false;
+  };
+  lm-math = fetchzip {
+    url = "http://www.gust.org.pl/projects/e-foundry/lm-math/download/latinmodern-math-1959.zip";
+    sha256 = "15l3lxjciyjmbh0q6jjvzz16ibk4ij79in9fs47qhrfr2wrddpvs";
+    stripRoot = false;
```
Comment (vbgl) [inline pkgs/tools/typesetting/satysfi/default.nix]: What’s the purpose of this `stripRoot` here?

**Review by vbgl — CHANGES_REQUESTED** [review]

File: pkgs/tools/typesetting/satysfi/default.nix (hunk 3 above)
Comment (mt-caret) [reply pkgs/tools/typesetting/satysfi/default.nix]: I guessed ocaml and ocamlbuild were only needed at build time. So this should all be in buildInputs?

**Comment by vbgl** [comment]
Does it work correctly? I tried the tests from https://github.com/gfngfn/SATySFi/tree/master/tests and got:

```
satysfi list.satyh math.satyh head.satyh first.saty -o output.pdf
 ---- ---- ---- ----
  target file: 'output.pdf'
! [Error] package file not found: dist/hash/default-font.satysfi-hash
make: *** [Makefile:4: all] Error 1
```

**Comment by mt-caret** [comment]
Yeah, that's a known issue (https://github.com/gfngfn/SATySFi/issues/17) (and probably one of the reasons why there's no official version out yet). Copying everything from ./dist to ~/.satysfi will fix it, but AFAIK that can't be done at build-time in Nix; I believe SATySFi currently looks for various files in hardcoded paths but that will change in the future.

**Comment by vbgl** [comment]
It’s easy to fix these hardcoded paths at configure-time. With this change, I could compile the `math1` example from the `tests` directory (other examples fail due to missing fonts).

**Review by vbgl — APPROVED** [review]

**Pull request merged**